concent_api/core/payments/storage.py

```diff
@@ -0,0 +1,145 @@
+                value    = int(ethereum_transaction.value),
+                v        = int(ethereum_transaction.v),
+                r        = int(ethereum_transaction.r),
+                s        = int(ethereum_transaction.s),
+                to       = encode_hex(ethereum_transaction.to.tobytes()),
+                data     = ethereum_transaction.data.tobytes(),
+            )
+            for ethereum_transaction in PendingEthereumTransaction.objects.all()
+        ]
+
+    @transaction.atomic(using='control')
+    def put_tx_and_inc_nonce(self, tx: Transaction) -> None:
+        """
+        Save a valid transaction and increase the nonce.
+        """
+        global_transaction_state = self._get_locked_global_transaction_state()
+
+        if global_transaction_state.nonce != tx.nonce:
+            raise Exception(
+                f'Transaction nonce does not match. '
+                f'Current={global_transaction_state.nonce}, '
+                f'Transaction={tx.nonce}'
+            )
+
+        pending_ethereum_transaction = PendingEthereumTransaction(
+            nonce       = tx.nonce,
+            gasprice    = tx.gasprice,
+            startgas    = tx.startgas,
+            value       = tx.value,
+            v           = tx.v,
+            r           = tx.r,
+            s           = tx.s,
+            data        = tx.data,
+            to          = tx.to,
+        )
+        pending_ethereum_transaction.full_clean()
+        pending_ethereum_transaction.save()
+
+        global_transaction_state.nonce += 1
+        global_transaction_state.full_clean()
+        global_transaction_state.save()
+
+    @transaction.atomic(using='control')
+    def remove_tx(self, nonce: int) -> None:
+        """
+        Remove the transaction after it's been confirmed and doesn't have
+        to be tracked anymore.
+        """
+
+        # This is called here only to lock GlobalTransactionState
+        self._get_locked_global_transaction_state()
+
+        try:
+            pending_ethereum_transaction = PendingEthereumTransaction.objects.get(nonce=nonce)
+            pending_ethereum_transaction.delete()
+            logger.info(f'Successfully removed PendingEthereumTransaction with nonce {nonce}.')
+        except PendingEthereumTransaction.DoesNotExist:
+            logger.error(f'Trying to remove PendingEthereumTransaction with nonce {nonce} but it does not exist.')
+            raise
+
+    @transaction.atomic(using='control')
+    def revert_last_tx(self) -> None:
+        """
+        Remove the last transaction that was added.
+        This shouldn't be ever called if everything is being used correctly,
+        i.e. we don't try to send invalid transactions.
+        """
+        global_transaction_state = self._get_locked_global_transaction_state()
+
+        try:
+            pending_ethereum_transaction = PendingEthereumTransaction.objects.get(
+                nonce=global_transaction_state.nonce - 1
+            )
```

Comment:
Your're not decreasing the nonce. This means that you can't for example call `revert_last_tx()` twice to revert two last transactions.

Is this intentional (maybe it's how SCI works)? Or is it a bug?

Reply:
My mistake, I thought `nonce` can never ever be used twice, but I have now checked that `JsonTransactionsStorage` is doing it.